extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/StartupBuildSteps.java

```diff
@@ -179,6 +180,27 @@ private void registerStartupObserver(ObserverRegistrationPhaseBuildItem observer
             ResultHandle containerHandle = mc.invokeStaticMethod(ARC_CONTAINER);
             ResultHandle beanHandle = mc.invokeInterfaceMethod(ARC_CONTAINER_BEAN, containerHandle,
                     mc.load(bean.getIdentifier()));
+
+            // if the [synthetic] bean is not active and is not injected in a user bean, skip obtaining the instance
+            // this means that an inactive bean that is injected into a user bean will end up with an error
```

Comment:
_"injected into a user bean"_ is IMO not accurate because the code only filters synthetic beans, i.e. class-based beans from extensions and libraries are considered "user beans".

Reply:
I think that's expected, actually. The point is that only synthetic beans may be inactive -- all class-based beans (and producer beans) are always active, so if they inject the inactive bean, that should be an error, always.

Reply:
I'll try to rewrite the comment.

Reply:
My point is that we should not call them "user beans" ;-).

> so if they inject the inactive bean, that should be an error, always.

That's weird. I mean, if an bean is injected into a non-synthetic bean we will initialize the bean during startup as normally and if the bean that injects this inactive bean is never created a user will never know there's a problem?

Reply:
I guess we could replace `!ip.getTargetBean().get().isSynthetic()` with `!ip.getTargetBean().get().canBeInactive()` and rename `isInjectedInUserBean` to `isInjectedInAlwaysActiveBean`. I guess that would reflect the intent better. We skip initialization here because that bean may also be inactive -- but that only applies to a subset of synthetic beans.

Reply:
> I guess we could replace `!ip.getTargetBean().get().isSynthetic()` with `!ip.getTargetBean().get().canBeInactive()` and rename `isInjectedInUserBean` to `isInjectedInAlwaysActiveBean`. I guess that would reflect the intent better.

:+1: 

> We skip initialization here because that bean may also be inactive -- but that only applies to a subset of synthetic beans.

Ok, so my previous comment was wrong. The check is only perfomed if the bean `A` is NOT injected in a non-synthetic bean. In other words, we only "deactivate" the bean if it's not injected somewhere?

Reply:
Yeah, sure. If the possibly-inactive synth bean is injected into an always active bean (which are all non-synth beans and some synth beans), we let the eager init happen, which will throw an exception. There's nothing else we can do.

If, however, the possibly-inactive synth bean is only injected into other possibly-inactive synth beans, we skip the eager init, because it might not be needed at all in the end.

Reply:
Done.

Reply:
> If, however, the possibly-inactive synth bean is only injected into other possibly-inactive synth beans, we skip the eager init, because it might not be needed at all in the end.

This is really weird. `@Startup` is not only about "eager init" of a bean but about executing _any logic_ during startup. Also the bean can be looked up programmatically.

I would naively expect that the synthetic observer for bean `A` (configured with `BeanConfigurator#startup()`) would be notified iff the bean is active. No matter if it is injected anywhere.

Reply:
Sure, but doing that would mean that we don't throw an exception in case the bean is inactive. Which we don't want. We want, as much as possible, to throw an exception eagerly if an inactive bean is used somewhere.

Reply:
I don't understand. The observer obtains the beans instance first. So it would check the inactivity when the bean instance is created and fail. At least that's what I see in the `BeanGenerator#implementCreateForSyntheticBean()`.

I'm sorry for my ignorance. Just trying to understand the logic and I might be missing something obvious.

Reply:
That's exactly right. So we let the observer create an instance with the expectation that it will fail.

We only short-circuit that, skipping eager init, if the bean is inactive and is only injected into other possibly inactive beans. In that case, we don't know if we should throw an error eagerly or not. In all other cases, we do know that.

Reply:
I think this is done, right?